fireplace/dsl/random_picker.py

```diff
@@ -1,79 +1,96 @@
 import random
 from hearthstone.enums import CardType, Race, Rarity
 from .lazynum import LazyValue
+from copy import deepcopy
+from .. import cards
 
 
 class RandomCardPicker(LazyValue):
 	"""
 	Store filters and generate a random card matching the filters on pick()
+	Constructor takes a single global set of filters and uses a default weighting of 1
+	Additional weighted filter sets can be added with add(), these will be merged with the global filters
 	"""
-	def __init__(self, *args, **filters):
-		self.args = args
-		self.filters = filters
+	def __init__(self, **filters):
+		self.weights = []
+		self.weightedfilters = []
+		self.globalfilter = filters
 		self.count = 1
-		self._cards = None
-		self.lazy_filters = False
-		for v in filters.values():
-			if isinstance(v, LazyValue):
-				self.lazy_filters = True
-				break
 
 	def __repr__(self):
-		return "%s(%r)" % (self.__class__.__name__, self.filters)
+		return "%s(%r)" % (self.__class__.__name__, self.globalfilter)
 
+	# select number of cards to fetch
 	def __mul__(self, other):
-		ret = self.__class__(*self.args, **self.filters)
+		ret = deepcopy(self)
 		ret.count = other
 		return ret
 
-	@property
-	def cards(self):
-		if self._cards is None:
-			self._cards = self._filter_cards(self.filters)
-		return self._cards
+	# add a filter set
+	def add(self, weight, **filters):
```

Comment:
So the other option is to push the weights straight into the constructor.

I don't have a clear idea of the Discover code you plan to write, but if you will know the weights ahead of time, I think it will be simpler to just push that into __init__ rather than mutating the picker on the fly.